---------------------------------------------------------------------------------------------------------------
# Modeling Summary Pipeline
---------------------------------------------------------------------------------------------------------------
## Content log:
1. Import Libraries
2. Load experiment results
3. Convert to Dataframe
4. Finding the Best Model
5. Loading final metrices of best model
6. Key Observations
---------------------------------------------------------------------------------------------------------------

# 01. Import Libraries

In [2]:
import json
import pandas as pd

# 02. Load experiment results

In [3]:
with open("../../models/experiment_results.json", "r") as f:
    results = json.load(f)
results

{'logistic_regression': {'accuracy': 0.7985150185622679,
  'precision': 0.5540110049232552,
  'recall': 0.5696843359142346,
  'f1_score': 0.5617383644105124,
  'roc_auc': 0.8152523642406756},
 'random_forest': {'accuracy': 0.8239622004724941,
  'precision': 0.7177700348432056,
  'recall': 0.3680762358546754,
  'f1_score': 0.48661417322834644,
  'roc_auc': 0.8332176666171248},
 'gradient_boosting': {'accuracy': 0.8400269996625043,
  'precision': 0.7400388726919339,
  'recall': 0.4535437760571769,
  'f1_score': 0.5624076809453471,
  'roc_auc': 0.8465571710946745},
 'xgboost': {'accuracy': 0.8373945325683428,
  'precision': 0.7333989178553861,
  'recall': 0.4440142942227516,
  'f1_score': 0.5531441291040623,
  'roc_auc': 0.8445759302086269},
 'best_model': 'gradient_boosting',
 'best_f1_score': 0.5624076809453471}

# 03. Convert to Dataframe

In [4]:
df = pd.DataFrame(results).T
df = df.drop(["best_model", "best_f1_score"], errors="ignore")
df

,accuracy,precision,recall,f1_score,roc_auc
logistic_regression,0.798515,0.554011,0.569684,0.561738,0.815252
random_forest,0.823962,0.71777,0.368076,0.486614,0.833218
gradient_boosting,0.840027,0.740039,0.453544,0.562408,0.846557
xgboost,0.837395,0.733399,0.444014,0.553144,0.844576


# 04. Finding the Best Model

In [5]:
best_model = results["best_model"]
best_f1 = results["best_f1_score"]
print("Best Model:", best_model)
print("Best F1 Score:", best_f1)

Best Model: gradient_boosting
Best F1 Score: 0.5624076809453471


# 05. Loading final metrices of best model

In [6]:
with open("../../models/final_metrics.json", "r") as f:
    final_metrics = json.load(f)
final_metrics

{'accuracy': 0.8400269996625043,
 'precision': 0.7400388726919339,
 'recall': 0.4535437760571769,
 'f1_score': 0.5624076809453471,
 'roc_auc': 0.8465571710946745}

# 06. Key Observations

MARKDOWN (WRITE THIS MANUALLY IN NOTEBOOK)
🔹 Summary
Multiple models evaluated: Logistic Regression, Random Forest, Gradient Boosting, XGBoost
Gradient Boosting selected as best model
🔹 Key Metrics
F1 Score ≈ 0.56
ROC-AUC ≈ 0.84
🔹 Insights
Model captures directional movement reasonably well
Higher precision than recall → conservative predictions
Suitable for financial decision support